src/server/pfs/server/driver.go

```diff
@@ -1085,84 +1081,103 @@ func (d *driver) propagateBranches(ctx context.Context, txnCtx *txncontext.Trans
+			inspectedCommit, err := d.inspectProcessingCommits(ctx, commitWithId, wait)
+			if err != nil {
+				return nil, err
+			}
+			return inspectedCommit.CommitInfo, nil
 		}
 	}
-	return commitInfo, nil
+	return commitWithId.CommitInfo, nil
 }
 
 // inspectProcessingCommits waits for the commit to be FINISHING or FINISHED.
-func (d *driver) inspectProcessingCommits(ctx context.Context, commitInfo *pfs.CommitInfo, wait pfs.CommitState) (*pfs.CommitInfo, error) {
-	var commitID pfsdb.CommitID
-	if err := d.txnEnv.WithReadContext(ctx, func(ctx context.Context, txnCtx *txncontext.TransactionContext) error {
-		var err error
-		commitID, err = pfsdb.GetCommitID(ctx, txnCtx.SqlTx, commitInfo.Commit)
-		if err != nil {
-			return err
-		}
-		return nil
-	}); err != nil {
-		return nil, err
-	}
+func (d *driver) inspectProcessingCommits(ctx context.Context, commitWithID *pfsdb.CommitWithID, wait pfs.CommitState) (*pfsdb.CommitWithID, error) {
 	// We only cancel the watcher if we detect the commit is the right state.
 	expectedErr := errors.New("commit is in the right state")
 	ctx, cancel := context.WithCancelCause(ctx)
 	defer cancel(nil)
-	if err := pfsdb.WatchCommit(ctx, d.env.DB, d.env.Listener, commitID,
+	if err := pfsdb.WatchCommit(ctx, d.env.DB, d.env.Listener, commitWithID.ID,
 		func(id pfsdb.CommitID, ci *pfs.CommitInfo) error {
 			switch wait {
 			case pfs.CommitState_FINISHING:
 				if ci.Finishing != nil {
-					commitInfo = ci
+					commitWithID.CommitInfo = ci
 					cancel(expectedErr)
 					return nil
 				}
 			case pfs.CommitState_FINISHED:
 				if ci.Finished != nil {
-					commitInfo = ci
+					commitWithID.CommitInfo = ci
 					cancel(expectedErr)
 					return nil
 				}
 			}
 			return nil
 		},
 		func(id pfsdb.CommitID) error {
-			return pfsserver.ErrCommitDeleted{Commit: commitInfo.Commit}
+			return pfsserver.ErrCommitDeleted{Commit: commitWithID.Commit}
 		},
 	); err != nil && !errors.Is(context.Cause(ctx), expectedErr) {
 		return nil, errors.Wrap(err, "inspect finishing or finished commit")
 	}
-	return commitInfo, nil
+	return commitWithID, nil
 }
 
-// resolveCommitWithAuth is like resolveCommit, but it does some pre-resolution checks like repo authorization.
-func (d *driver) resolveCommitWithAuth(ctx context.Context, commit *pfs.Commit) (*pfs.CommitInfo, error) {
+// resolveCommitWithIdAndAuth is like resolveCommitWithID, but it does some pre-resolution checks like repo authorization.
+func (d *driver) resolveCommitWithIdAndAuth(ctx context.Context, commit *pfs.Commit) (*pfsdb.CommitWithID, error) {
 	if commit.Repo.Name == fileSetsRepo {
 		cinfo := &pfs.CommitInfo{
 			Commit:      commit,
 			Description: "FileSet - Virtual Commit",
 			Finished:    &timestamppb.Timestamp{}, // it's always been finished. How did you get the id if it wasn't finished?
 		}
-		return cinfo, nil
+		return &pfsdb.CommitWithID{
+			ID:         0, // This doesn't seem like the right thing to do, but here we are.
```

Comment:
Yeah, this isn't great, but removing the fileset repository is on the roadmap. We should have a stronger guarantee that the IDs are set after this.